ocaml/testsuite/tests/typing-modal-kinds/basics.ml

```diff
@@ -0,0 +1,559 @@
+                                                 ^
+Error: This value escapes its region
+|}]
+
+let function_escape = let local_ x : int -> int = fun y -> y in x
+
+[%%expect{|
+Line 1, characters 64-65:
+1 | let function_escape = let local_ x : int -> int = fun y -> y in x
+                                                                    ^
+Error: This value escapes its region
+|}]
+
+let function_list_escape =
+  let local_ x : (int -> int) list = [(fun y -> y); fun z -> z + 1] in x
+
+[%%expect{|
+Line 2, characters 71-72:
+2 |   let local_ x : (int -> int) list = [(fun y -> y); fun z -> z + 1] in x
+                                                                           ^
+Error: This value escapes its region
+|}]
+
+let string_duplicate = let once_ x : string = "hello" in Fun.id x
+
+(* CR layouts v2.8: this should succeed *)
+[%%expect{|
+Line 1, characters 64-65:
+1 | let string_duplicate = let once_ x : string = "hello" in Fun.id x
+                                                                    ^
+Error: Found a once value where a many value was expected
+|}]
+
+let int_duplicate = let once_ x : int = 5 in Fun.id x
+
+[%%expect{|
+val int_duplicate : int = 5
+|}]
+
+let string_list_duplicate = let once_ x : string list = ["hi";"bye"] in Fun.id x
+
+(* CR layouts v2.8: this should succeed *)
+[%%expect{|
+Line 1, characters 79-80:
+1 | let string_list_duplicate = let once_ x : string list = ["hi";"bye"] in Fun.id x
+                                                                                   ^
+Error: Found a once value where a many value was expected
+|}]
+
+let int_list_duplicate = let once_ x : int list = [4;5] in Fun.id x
+
+(* CR layouts v2.8: this should succeed *)
+[%%expect{|
+Line 1, characters 66-67:
+1 | let int_list_duplicate = let once_ x : int list = [4;5] in Fun.id x
+                                                                      ^
+Error: Found a once value where a many value was expected
+|}]
+
+let hidden_string_duplicate =
+  let once_ x : Hidden_string.t = Hidden_string.hide "hello" in Fun.id x
+
+[%%expect{|
+Line 2, characters 71-72:
+2 |   let once_ x : Hidden_string.t = Hidden_string.hide "hello" in Fun.id x
+                                                                           ^
+Error: Found a once value where a many value was expected
+|}]
+
+let hidden_int_duplicate =
+  let once_ x : Hidden_int.t = Hidden_int.hide 42 in Fun.id x
+
+[%%expect{|
+val hidden_int_duplicate : Hidden_int.t = <abstr>
+|}]
+
+let hidden_string_list_duplicate =
+  let once_ x : Hidden_string.t list =
+    [Hidden_string.hide "hi"; Hidden_string.hide "bye"]
+  in Fun.id x
+
+[%%expect{|
+Line 4, characters 12-13:
+4 |   in Fun.id x
+                ^
+Error: Found a once value where a many value was expected
+|}]
+
+let hidden_int_list_duplicate =
+  let once_ x : Hidden_int.t list =
+    [Hidden_int.hide 2; Hidden_int.hide 3]
+  in Fun.id x
+
+(* CR layouts v2.8: this should succeed *)
```

Comment:
Thanks for the comment - I suppose the scope of this PR is only about immediate, and therefore "arrow types crosses linearity" is left as future PRs? I'm fine with that - just confirming (because the title of this PR suggests otherwise)

Reply:
I'm not sure I agree about the comment about the title (this really does use modal kinds to control mode-crossing), but yes I plan on fixing that stuff in future work.